### ISSUE TITLE
Implement a synchronization server

### ISSUE THREAD

**Issue by djmitche** [djmitche/taskchampion#8]
Something that exposes an HTTP API against which `task sync` can run

**Comment by djmitche** [comment]
We want to be able to "sync" locally, too, so implement this with a trait that can be boxed so that `task sync` can talk via HTTP or locally by just handing it a different `dyn TaskServer`.

**Comment by djmitche** [comment]
The current Server interface has a "username" which suggests that the server can store multiple users' data, but doesn't do any kind of authentication.

One of the nice features of TaskWarrior's taskd, at least as it's run by freecinc, is that it doesn't have names of users.  Let's keep that up.

Rather than thinking of a "user", let's think of a particular "task history" starting at the empty version 0.  That history is the context for the monotonic version history.  So maybe creation of a new DB should tag that DB with its identity in the form of a v4 uuid.

I think we need to consider how setup occurs. Likely the most common case will be people setting up with a "local" sync operation and transitioning to a server, so we should support that mode.  So it should be possible to push a bunch of versions to an empty account on a server.

But, we'll want to avoid the case where two servers have the same set of data, with the same identity uuid, and start to diverge.  Then version N will be different on those two servers and all hell will break loose.  So perhaps versions aren't monotonic integers, but are a content-addressed linked list, each with the identity of the accepting server.

I think it's important to work these details out now as part of the MVP, even if the implementations cut some corners.

So that's adding some hashes and HMACs, and I think we'll need a bit of asymmetric-key encryption too.  Once it's clear what's required, let's get those crypto primitives -- based on an existing library of course -- nicely encapsulated.

**Comment by djmitche** [comment]
## Client-Server Protocol

This protocol uses the term client instead of replica. Multiple replicas are
considered as the same "client".  Clients are identified by clientId and
authenticated via asymmetric key signatures.

The server's primary responsibility is to build a single linked list of
versions for the client.  Primary responsibility for protecting the data in
those versions rests with the client (outside of the protocol), but read
operations are also authenticated with signatures to prevent abuse.

Client-server communication is via request/response transactions.

### Storage

Client storage:
  Information:
    - clientId: uuid
    - pubKey: key
    - privKey: key (shared between replicas, not on server)
  TaskDB

Server Storage:
  - Clients:
    - clientId (pk)
    - pubKey
    - latestVersion: id of latest accepted version, or 0's for new client
  - Versions:
    - id: sha512(acceptedNonce || clientId || parentVersion || historySegment) (pk)
    - acceptedNonce: uuid (used for entropy in id's)
    - clientId: uuid
    - parentVersion: sha512 (id of parent version, or 0's) (indexed)
    - historySegment: bytes (containing a list of operations; opaque to server)
    - signature: sign(client privKey, clientId || parentVersion || historySegment)

### Invariants

1. Given a client c, c.latestVersion is 0's or exists in the set of
   Versions.
1. Given a version v, v.clientId exists in the set of Clients.
1. Given a version v, v.signature verifies with the client's pubKey
1. Given versions v and pv with v.parentVersion == pv.id, v.clientId ==
   pv.clientId
1. Given versions v1 and v2 with v1.id != v2.id and v1.parentVersion != 0's,
   v1.parentVersion != v2.parentVersion.  In other words, versions do not
   branch.

Note that versions form a linked list beginning with the version stored in he
client.  This linked list need not continue back to a version with
v.parentVersion = 0's.  It may end at any point when v.parentVersion is not
found in the set of Versions.

### AddVersion Transaction

AddVersion request:
  - clientId: uuid
  - parentVersion: sha512
  - historySegment: bytes
  - signature

AddVersion response:
  - id

Handling an AddVersion request
  - look up req.clientId
  - verify signature: verify(client pubKey, req.clientId || req.parentVersion || req.historySegment)
  - if client.latestVersion is NOT 0's
    - if v.parentVersion != client.latestVersion:
      - return rejection with client.latestVersion
  - generate acceptedNonce
  - generate id
  - store version, update client.latestVersion
  - return id

### GetChildVersion Transaction

GetChildVersion request:
  - clientId
  - id
  - signature: sign(client privKey, req.clientId || req.id)

GetChildVersion response:
  - (all version fields)

Handling a GetChildVersion request
  - look up req.clientId
  - verify signature: verify(client pubKey, req.clientId || req.id)
  - look up version by version.parent = req.id
  - verify version.clientId == req.clientId
  - return version

### Switching Servers

A client may switch to a new server by supplying that server with its clientId
and pubKey, then adding a new version on that server.  However, if replicas
comprising that client are not synchronized before this occurs, then their
task histories may diverge, requiring re-initialization of some replicas.

The expected use-case for switching servers is to switch from a single replica
communicating with a local "server" to communicating to a remote server, after
which time additional replicas might be added.  In this use-case, there is
only one replica, so divergence is not a concern.

**Comment by djmitche** [comment]
That will be the "core' of the protocol.  There's lots of stuff to add there, and notably snapshots will be critical to adding a new replica to a client.

**Comment by djmitche** [comment]
I've revamped the server interface based on the above, and implemented a `LocalServer` that stores server data locally.  Aside from the automated testing, I've tested this manually and it seems to work, even sync'ing changes made concurrently in two different replicas.  Next steps:

 * [ ] set up an http server in sync-server/
 * [ ] make a server backend trait (similar to `taskchampion::server:Server`, but with `client_id` as a parameter to some methods, and also methods to get/add clients)
 * [ ] add `/add-version` and `/get-child-version` endpoints that verify signatures and call the backend
 * [ ] build a `RemoteServer` class that calls those methods.  This might be a good time to add `taskchampion::sync` that can handle the intermediate signing bits instead of putting those in TaskDB.
 * [ ] add enough temporary stuff to get things set up without requiring config files (maybe a CLI to add clients on the server and env vars on the cli?)
 * [ ] update docs based on comments above
